File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/widgets/WidgetChooseTimeRange.java
package com.sup.dev.android.views.widgets;

import android.support.annotation.StringRes;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.sup.dev.android.androiddevsup.R;
import com.sup.dev.android.tools.ToolsResources;
import com.sup.dev.android.tools.ToolsView;
import com.sup.dev.android.views.screens.SWidget;
import com.sup.dev.java.classes.callbacks.simple.Callback1;
import com.sup.dev.java.tools.ToolsDate;

public class WidgetChooseTimeRange extends Widget {

    private final Button vStart;
    private final Button vEnd;
    private final Button vCancel;
    private final Button vEnter;
    private final TextView vTitle;

    private boolean autoHideOnEnter = true;

    public WidgetChooseTimeRange() {
        super(R.layout.widget_choose_time_range);

        vStart = view.findViewById(R.id.start);
        vEnd = view.findViewById(R.id.end);
        vCancel = view.findViewById(R.id.cancel);
        vEnter = view.findViewById(R.id.enter);
        vTitle = view.findViewById(R.id.title);

        vCancel.setVisibility(View.GONE);
        vEnter.setVisibility(View.GONE);

        vStart.setOnClickListener(v -> new WidgetChooseTime()
                .setOnCancel(vCancel.getText().toString())
                .setOnEnter(vEnter.getText().toString(), (dialog, h, m) -> setTimeStart(h, m))
                .asDialogShow());

        vEnd.setOnClickListener(v -> new WidgetChooseTime()
                .setOnCancel(vCancel.getText().toString())
                .setOnEnter(vEnter.getText().toString(), (dialog, h, m) -> setTimeEnd(h, m))
                .asDialogShow());

        setTimeStart(12, 0);
        setTimeEnd(24, 0);
    }

    @Override
    public void onShow() {
        super.onShow();

        ToolsView.setTextOrGone(vTitle, vTitle.getText());
        if(viewWrapper instanceof SWidget){
            vTitle.setVisibility(View.GONE);
            ((SWidget)viewWrapper).setTitle(vTitle.getText().toString());
        }
    }

    //
    //  Setters
    //

    public WidgetChooseTimeRange setTitle(@StringRes int title) {
        return setTitle(ToolsResources.getString(title));
    }

    public WidgetChooseTimeRange setTitle(String title) {
        ToolsView.setTextOrGone(vTitle, title);
        return this;
    }

    public WidgetChooseTimeRange setTimeStart(int h, int m) {
        vStart.setText(ToolsDate.timeToString(h, m));
        return this;
    }

    public WidgetChooseTimeRange setTimeEnd(int h, int m) {
        vEnd.setText(ToolsDate.timeToString(h, m));
        return this;
    }

    public WidgetChooseTimeRange setOnEnter(@StringRes int s) {
        return setOnEnter(ToolsResources.getString(s));
    }

    public WidgetChooseTimeRange setOnEnter(String s) {
        return setOnEnter(s, null);
    }

    public WidgetChooseTimeRange setOnEnter(@StringRes int s, Callback1<WidgetChooseTimeRange> onEnter) {
        return setOnEnter(ToolsResources.getString(s), onEnter);
    }

    public WidgetChooseTimeRange setOnEnter(String s, Callback1<WidgetChooseTimeRange> onEnter) {
        ToolsView.setTextOrGone(vEnter, s);
        vEnter.setOnClickListener(v -> {
            if (autoHideOnEnter) hide();
            else setEnabled(false);
            if (onEnter != null) onEnter.callback(this);
        });
        return this;
    }

    public WidgetChooseTimeRange setAutoHideOnEnter(boolean autoHideOnEnter) {
        this.autoHideOnEnter = autoHideOnEnter;
        return this;
    }

    public WidgetChooseTimeRange setOnCancel(@StringRes int s) {
        return setOnCancel(ToolsResources.getString(s), null);
    }

    public WidgetChooseTimeRange setOnCancel(String s) {
        return setOnCancel(s, null);
    }

    public WidgetChooseTimeRange setOnCancel(Callback1<WidgetChooseTimeRange> onCancel) {
        return setOnCancel(null, onCancel);
    }

    public WidgetChooseTimeRange setOnCancel(@StringRes int s, Callback1<WidgetChooseTimeRange> onCancel) {
        return setOnCancel(ToolsResources.getString(s), onCancel);
    }

    public WidgetChooseTimeRange setOnCancel(String s, Callback1<WidgetChooseTimeRange> onCancel) {
        super.setOnHide(b -> {
            if (onCancel != null) onCancel.callback(this);
        });
        ToolsView.setTextOrGone(vCancel, s);
        vCancel.setOnClickListener(v -> {
            hide();
            if (onCancel != null) onCancel.callback(this);
        });
        return this;
    }

    public WidgetChooseTimeRange setEnabled(boolean enabled) {
        super.setEnabled(enabled);
        return this;
    }


}

File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/widgets/WidgetChooseTimeRange.kt
package com.sup.dev.android.views.widgets

import android.support.annotation.StringRes
import android.view.View
import android.widget.Button
import com.sup.dev.android.R
import com.sup.dev.android.app.SupAndroid
import com.sup.dev.android.tools.ToolsResources
import com.sup.dev.android.tools.ToolsView
import com.sup.dev.java.tools.ToolsDate


class WidgetChooseTimeRange : Widget(R.layout.widget_choose_time_range) {

    private val vStart: Button = view.findViewById(R.id.vStart)
    private val vEnd: Button = view.findViewById(R.id.vEnd)
    private val vCancel: Button = view.findViewById(R.id.vCancel)
    private val vEnter: Button = view.findViewById(R.id.vEnter)

    private var h1: Int = 0
    private var m1: Int = 0
    private var h2: Int = 0
    private var m2: Int = 0

    private var autoHideOnEnter = true

    init {

        isUseMoreScreenSpace = true
        vCancel.visibility = View.GONE
        vEnter.visibility = View.GONE

        vStart.setOnClickListener { v ->
            WidgetChooseTime()
                .setOnCancel(SupAndroid.TEXT_APP_CANCEL)
                .setOnEnter(SupAndroid.TEXT_APP_CHOOSE) { dialog, h, m -> setTimeStart(h, m) }
                .asSheetShow()
        }

        vEnd.setOnClickListener { v ->
            WidgetChooseTime()
                .setOnCancel(SupAndroid.TEXT_APP_CANCEL)
                .setOnEnter(SupAndroid.TEXT_APP_CHOOSE) { dialog, h, m -> setTimeEnd(h, m) }
                .asSheetShow()
        }

        setTimeStart(12, 0)
        setTimeEnd(24, 0)
    }

    //
    //  Setters
    //

    fun setTimeStart(h: Int, m: Int): WidgetChooseTimeRange {
        h1 = h
        m1 = m
        vStart.text = ToolsDate.timeToString(h, m)
        return this
    }

    fun setTimeEnd(h: Int, m: Int): WidgetChooseTimeRange {
        h2 = h
        m2 = m
        vEnd.text = ToolsDate.timeToString(h, m)
        return this
    }

    fun setOnEnter(@StringRes s: Int): WidgetChooseTimeRange {
        return setOnEnter(ToolsResources.s(s))
    }

    fun setOnEnter(@StringRes s: Int, onEnter: (WidgetChooseTimeRange, Int, Int, Int, Int) -> Unit): WidgetChooseTimeRange {
        return setOnEnter(ToolsResources.s(s), onEnter)
    }

    @JvmOverloads
    fun setOnEnter(
        s: String?,
        onEnter: (WidgetChooseTimeRange, Int, Int, Int, Int) -> Unit = { w, x, y, xx, yy -> }
    ): WidgetChooseTimeRange {
        ToolsView.setTextOrGone(vEnter, s)
        vEnter.setOnClickListener { v ->
            if (autoHideOnEnter)
                hide()
            else
                setEnabled(false)
            onEnter.invoke(this, h1, m1, h2, m2)
        }
        return this
    }

    fun setAutoHideOnEnter(autoHideOnEnter: Boolean): WidgetChooseTimeRange {
        this.autoHideOnEnter = autoHideOnEnter
        return this
    }

    fun setOnCancel(onCancel: (WidgetChooseTimeRange) -> Unit = {}): WidgetChooseTimeRange {
        return setOnCancel(null, onCancel)
    }

    fun setOnCancel(@StringRes s: Int, onCancel: (WidgetChooseTimeRange) -> Unit): WidgetChooseTimeRange {
        return setOnCancel(ToolsResources.s(s), onCancel)
    }

    @JvmOverloads
    fun setOnCancel(s: String?, onCancel: (WidgetChooseTimeRange) -> Unit = {}): WidgetChooseTimeRange {
        super.setOnHide { b -> onCancel.invoke(this) }
        ToolsView.setTextOrGone(vCancel, s)
        vCancel.setOnClickListener { v ->
            hide()
            onCancel.invoke(this)
        }
        return this
    }

}
